sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,539 @@
+ * <p>
+ * <p>To configure the MongoDB source, you have to provide the connection URI, the database name
+ * and the collection name. The following example illustrates various options for configuring the
+ * source:</p>
+ * <p>
+ * <pre>{@code
+ *
+ * pipeline.apply(MongoDbIO.read()
+ *   .withUri("mongodb://localhost:27017")
+ *   .withDatabase("my-database")
+ *   .withCollection("my-collection"))
+ *   // above three are required configuration, returns PCollection<String>
+ *
+ *   // rest of the settings are optional
+ *
+ * }</pre>
+ * <p>
+ * <p>The source also accepts an optional configuration: {@code withFilter()} allows you to
+ * define a JSON filter to get subset of data.</p>
+ * <p>
+ * <h3>Writing to MongoDB</h3>
+ * <p>
+ * <p>MongoDB sink supports writing of Document (as JSON String) in a MongoDB.</p>
+ * <p>
+ * <p>To configure a MongoDB sink, you must specify a connection {@code URI}, a {@code Database}
+ * name, a {@code Collection} name. For instance:</p>
+ * <p>
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...)
+ *   .apply(MongoDbIO.write()
+ *     .withUri("mongodb://localhost:27017")
+ *     .withDatabase("my-database")
+ *     .withCollection("my-collection")
+ *     .withNumSplits(30))
+ *
+ * }</pre>
+ */
+// TODO instead of JSON String, does it make sense to populate the PCollection with BSON Document or
+//  DBObject ??
+public class MongoDbIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbIO.class);
+
+  /** Read data from MongoDB. */
+  public static Read read() {
+    return new Read(new BoundedMongoDbSource(null, null, null, null, 0));
+  }
+
+  /** Write data to MongoDB. */
+  public static Write write() {
+    return new Write(new Write.MongoDbWriter(null, null, null));
+  }
+
+  private MongoDbIO() {
+  }
+
+  /**
+   * A {@link PTransform} to read data from MongoDB.
+   */
+  public static class Read extends PTransform<PBegin, PCollection<String>> {
+
+    public Read withUri(String uri) {
+      return new Read(source.withUri(uri));
+    }
+
+    public Read withDatabase(String database) {
+      return new Read(source.withDatabase(database));
+    }
+
+    public Read withCollection(String collection) {
+      return new Read(source.withCollection(collection));
+    }
+
+    public Read withFilter(String filter) {
+      return new Read(source.withFilter(filter));
+    }
+
+    public Read withNumSplits(int numSplits) {
+      return new Read(source.withNumSplits(numSplits));
+    }
+
+    private final BoundedMongoDbSource source;
+
+    private Read(BoundedMongoDbSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public PCollection<String> apply(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(createSource()));
+    }
+
+    /**
+     * Creates a {@link BoundedSource} with the configuration in {@link Read}.
+     */
+    @VisibleForTesting
+    BoundedSource<String> createSource() {
```

Comment:
Should probably be called getSource(), or just make the variable itself @VisibleForTesting.